CI: upgrade to Ubuntu 22.04

In #5407 we pinned to 20.04 because of issues such as missing `libclang1-9` in [this CI run](https://github.com/bytecodealliance/wasmtime/actions/runs/3659854687/jobs/6186416352) with an Ubuntu 22.04 image.

We should resolve all issues that appear on 22.04 and upgrade to the new version when possible.

After https://github.com/bytecodealliance/wasmtime/pull/5410 the only 20.04 builder is the wasi-nn builder, and @abrown [this is a failure on the 22.04 builder](https://github.com/bytecodealliance/wasmtime/actions/runs/3659854687/jobs/6186417666) which gives:

```
Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).
OK
deb https://apt.repos.intel.com/openvino/2022 jammy main

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Get:1 https://packages.microsoft.com/ubuntu/22.04/prod jammy InRelease [10.5 kB]
Hit:2 http://azure.archive.ubuntu.com/ubuntu jammy InRelease
Get:3 http://azure.archive.ubuntu.com/ubuntu jammy-updates InRelease [114 kB]
Get:4 http://azure.archive.ubuntu.com/ubuntu jammy-backports InRelease [99.8 kB]
Get:5 http://azure.archive.ubuntu.com/ubuntu jammy-security InRelease [110 kB]
Get:6 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main arm64 Packages [16.7 kB]
Get:7 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main armhf Packages [10.6 kB]
Get:8 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main amd64 Packages [60.1 kB]
Get:9 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [758 kB]
Get:10 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main Translation-en [169 kB]
Hit:11 https://ppa.launchpadcontent.net/ubuntu-toolchain-r/test/ubuntu jammy InRelease
Get:12 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 c-n-f Metadata [11.4 kB]
Get:13 http://azure.archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [488 kB]
Get:14 http://azure.archive.ubuntu.com/ubuntu jammy-updates/restricted Translation-en [74.9 kB]
Get:15 http://azure.archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [761 kB]
Get:16 http://azure.archive.ubuntu.com/ubuntu jammy-updates/universe Translation-en [128 kB]
Get:17 http://azure.archive.ubuntu.com/ubuntu jammy-updates/universe amd64 c-n-f Metadata [14.0 kB]
Get:18 http://azure.archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [7300 B]
Get:19 http://azure.archive.ubuntu.com/ubuntu jammy-backports/main amd64 Packages [3324 B]
Get:20 http://azure.archive.ubuntu.com/ubuntu jammy-backports/main Translation-en [1580 B]
Get:21 http://azure.archive.ubuntu.com/ubuntu jammy-backports/main amd64 c-n-f Metadata [272 B]
Get:22 http://azure.archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [6740 B]
Get:23 http://azure.archive.ubuntu.com/ubuntu jammy-backports/universe amd64 c-n-f Metadata [348 B]
Get:24 http://azure.archive.ubuntu.com/ubuntu jammy-security/main amd64 Packages [525 kB]
Get:25 http://azure.archive.ubuntu.com/ubuntu jammy-security/main Translation-en [113 kB]
Get:26 http://azure.archive.ubuntu.com/ubuntu jammy-security/main amd64 c-n-f Metadata [7352 B]
Get:27 http://azure.archive.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [452 kB]
Get:28 http://azure.archive.ubuntu.com/ubuntu jammy-security/restricted Translation-en [69.2 kB]
Get:29 http://azure.archive.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [622 kB]
Get:30 http://azure.archive.ubuntu.com/ubuntu jammy-security/universe Translation-en [82.7 kB]
Get:31 http://azure.archive.ubuntu.com/ubuntu jammy-security/universe amd64 c-n-f Metadata [11.0 kB]
Ign:32 https://apt.repos.intel.com/openvino/2022 jammy InRelease
Err:33 https://apt.repos.intel.com/openvino/2022 jammy Release
  404  Not Found [IP: 184.50.211.30 443]
Reading package lists...
E: The repository 'https://apt.repos.intel.com/openvino/2022 jammy Release' does not have a Release file.
Error: Process completed with exit code 100.
```

Upgrading to 22.04 is not of the utmost urgency at all, but might be good to forward to folks at Intel I think hosting openvino there?

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -466,14 +466,17 @@ jobs:
     needs: determine
     if: needs.determine.outputs.run-full
     name: Test wasi-nn module
-    runs-on: ubuntu-20.04 # TODO: remove pin when fixed (#5408)
+    runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
         with:
           submodules: true
       - uses: ./.github/actions/install-rust
       - run: rustup target add wasm32-wasi
-      - uses: abrown/install-openvino-action@v3
+      - uses: abrown/install-openvino-action@50a2ab733ef4b504dd7de3ac45c36eae7f9bc228
+        with:
+          version: 2022.3.0
+          apt: true
       - run: ./ci/run-wasi-nn-example.sh
         env:
           RUST_BACKTRACE: 1

--- a/ci/run-wasi-nn-example.sh
+++ b/ci/run-wasi-nn-example.sh
@@ -1,10 +1,10 @@
 #!/bin/bash
 
-# The following script demonstrates how to execute a machine learning inference using the wasi-nn module optionally
-# compiled into Wasmtime. Calling it will download the necessary model and tensor files stored separately in $FIXTURE
-# into $TMP_DIR (optionally pass a directory with existing files as the first argument to re-try the script). Then,
-# it will compile the example code in crates/wasi-nn/tests/example into a Wasm file that is subsequently
-# executed with the Wasmtime CLI.
+# The following script demonstrates how to execute a machine learning inference using the wasi-nn
+# module optionally compiled into Wasmtime. Calling it will download the necessary model and tensor
+# files stored separately in $FIXTURE into $TMP_DIR (optionally pass a directory with existing files
+# as the first argument to re-try the script). Then, it will compile the example code in
+# crates/wasi-nn/tests/example into a Wasm file that is subsequently executed with the Wasmtime CLI.
 set -e
 WASMTIME_DIR=$(dirname "$0" | xargs dirname)
 FIXTURE=https://github.com/intel/openvino-rs/raw/main/crates/openvino/tests/fixtures/mobilenet
@@ -18,12 +18,8 @@ else
     REMOVE_TMP_DIR=0
 fi
 
-# Inform the environment of OpenVINO library locations. Then we use OPENVINO_INSTALL_DIR below to avoid building all of
-# OpenVINO from source (quite slow).
-source /opt/intel/openvino_2022/setupvars.sh
-
 # Build Wasmtime with wasi-nn enabled; we attempt this first to avoid extra work if the build fails.
-OPENVINO_INSTALL_DIR=/opt/intel/openvino_2022 cargo build -p wasmtime-cli --features wasi-nn
+cargo build -p wasmtime-cli --features wasi-nn
 
 # Download all necessary test fixtures to the temporary directory.
 wget --no-clobber $FIXTURE/mobilenet.bin --output-document=$TMP_DIR/model.bin

--- a/crates/wasi-nn/Cargo.toml
+++ b/crates/wasi-nn/Cargo.toml
@@ -17,7 +17,7 @@ anyhow = { workspace = true }
 wiggle = { workspace = true }
 
 # These dependencies are necessary for the wasi-nn implementation:
-openvino = { version = "0.4.2", features = ["runtime-linking"] }
+openvino = { version = "0.5.0", features = ["runtime-linking"] }
 thiserror = { workspace = true }
 
 [build-dependencies]

--- a/crates/wasi-nn/src/openvino.rs
+++ b/crates/wasi-nn/src/openvino.rs
@@ -119,7 +119,7 @@ impl BackendExecutionContext for OpenvinoExecutionContext {
 
     fn get_output(&mut self, index: u32, destination: &mut [u8]) -> Result<u32, BackendError> {
         let output_name = self.0.get_output_name(index as usize)?;
-        let mut blob = self.1.get_blob(&output_name)?;
+        let blob = self.1.get_blob(&output_name)?;
         let blob_size = blob.byte_len()?;
         if blob_size > destination.len() {
             return Err(BackendError::NotEnoughMemory(blob_size));

--- a/supply-chain/audits.toml
+++ b/supply-chain/audits.toml
@@ -596,6 +596,11 @@ notes = """
 I am the author of most of these changes.
 """
 
+[[audits.openvino]]
+who = "Andrew Brown <andrew.brown@intel.com>"
+criteria = "safe-to-deploy"
+delta = "0.4.2 -> 0.5.0"
+
 [[audits.openvino-finder]]
 who = "Matthew Tamayo-Rios <matthew@geekbeast.com>"
 criteria = "safe-to-deploy"
@@ -604,6 +609,11 @@ notes = """
 Only updates to Cargo file for versioning.
 """
 
+[[audits.openvino-finder]]
+who = "Andrew Brown <andrew.brown@intel.com>"
+criteria = "safe-to-deploy"
+delta = "0.4.2 -> 0.5.0"
+
 [[audits.openvino-sys]]
 who = "Matthew Tamayo-Rios <matthew@geekbeast.com>"
 criteria = "safe-to-deploy"
@@ -611,6 +621,11 @@ delta = "0.4.1 -> 0.4.2"
 notes = """
 Only updates to tests to use new rust functions for mut pointers.
 """
+
+[[audits.openvino-sys]]
+who = "Andrew Brown <andrew.brown@intel.com>"
+criteria = "safe-to-deploy"
+delta = "0.4.2 -> 0.5.0"
 
 [[audits.peeking_take_while]]
 who = "Nick Fitzgerald <fitzgen@gmail.com>"

